src/base/bittorrent/session.cpp

```diff
@@ -1887,8 +1889,58 @@ bool Session::addTorrent_impl(CreateTorrentParams params, const MagnetUri &magne
 
         p = magnetUri.addTorrentParams();
     }
-    else if (torrentInfo.isValid()) {
-        if (!params.restored) {
+    else {
+        if (!torrentInfo.isValid()) {
+            // We can have an invalid torrentInfo when there isn't a matching
+            // .torrent file to the .fastresume we loaded. Possibly from a
+            // failed upgrade.
+            return false;
+        }
+
+        hash = torrentInfo.hash();
+    }
+
+    // We should not add the torrent if it is already
+    // processed or is pending to add to session
+    if (m_addingTorrents.contains(hash) || m_loadedMetadata.contains(hash))
+        return false;
+
+    TorrentHandle *const torrent = m_torrents.value(hash);
+    if (torrent) {  // a duplicate torrent is added
+        if (torrent->isPrivate() || (!fromMagnetUri && torrentInfo.isPrivate()))
+            return false;
+
+        // merge trackers and web seeds
+        torrent->addTrackers(fromMagnetUri ? magnetUri.trackers() : torrentInfo.trackers());
+        torrent->addUrlSeeds(fromMagnetUri ? magnetUri.urlSeeds() : torrentInfo.urlSeeds());
+        return true;
+    }
+
+    // Record if .fastresume file contains fetched metadata
+    // In the case of Magnet metadata isn't fetched yet, we still need
+    // to setup the default parameters.
+    bool hasCompleteFastresume = false;
+
+    if (!fromMagnetUri) {
+        if (params.restored) {  // load from existing fastresume
+#if (LIBTORRENT_VERSION_NUM < 10200)
+            p.resume_data = std::vector<char> {fastresumeData.constData()
+                , (fastresumeData.constData() + fastresumeData.size())};
+            p.flags |= lt::add_torrent_params::flag_use_resume_save_path;
+
+            // we still setup the default parameters and
+            // let libtorrent handle the parameter merging
+            hasCompleteFastresume = false;
+#else
+            lt::error_code ec;
+            p = lt::read_resume_data(fastresumeData, ec);
+
+            // fastresume will only contain "qBt-*" entries when metadata
+            // hasn't been fetched
```

Comment:
You still have incorrect comments. Apparently you still don't understand it correctly.
We are under `!fromMagnetUri` condition here that means we obviously have metadata (from torrent file). Since this method stops execution if both magnetUri and torrentInfo are invalid.
I think we can have incomplete fastresume data at this point only if the file is corrupted for some reason.
In any case, I wouldn't make a one-to-one correspondence between metadata and fastresume data in the comments, otherwise they might be misleading.

Reply:
>We are under !fromMagnetUri condition here that means we obviously have metadata (from torrent file).

Please explain to me why you claim there is metadata but there is no .torrent file as in the fake hash case?
.torrent file == metadata is what I read from libtorrent doc.

Reply:
>...this method stops execution if both magnetUri and torrentInfo are invalid.

What "fake hash case" do you mean?
This method accepts either torrent info or magnet uri. When you add torrent via magnet/hash, magnet uri is valid here (the same is true when you restore torrent previously added via magnet/hash until its metadata loaded). When you add/restore torrent that has metadata then magnetUri is invalid here and torrentInfo should be valid (otherwise the method stops its execution).

Reply:
I've updated the comments.